packages/sanity/src/structure/components/requestPermissionDialog/RequestPermissionDialog.tsx

```diff
@@ -0,0 +1,184 @@
+import {useTelemetry} from '@sanity/telemetry/react'
+import {Box, Card, DialogProvider, Flex, Stack, Text, TextInput, useToast} from '@sanity/ui'
+import {useId, useMemo, useState} from 'react'
+import {useObservable} from 'react-rx'
+import {catchError, map, type Observable, of, startWith} from 'rxjs'
+import {type Role, useClient, useProjectId, useTranslation, useZIndex} from 'sanity'
+import {styled} from 'styled-components'
+
+import {Dialog} from '../../../ui-components'
+import {structureLocaleNamespace} from '../../i18n'
+import {AskToEditRequestSent} from './__telemetry__/RequestPermissionDialog.telemetry'
+import {type AccessRequest} from './useRoleRequestsStatus'
+
+const MAX_NOTE_LENGTH = 150
+
+/** @internal */
+export const DialogBody = styled(Box)`
+  box-sizing: border-box;
+`
+
+/** @internal */
+export const LoadingContainer = styled(Flex).attrs({
+  align: 'center',
+  direction: 'column',
+  justify: 'center',
+})`
+  height: 110px;
+`
+
+/** @internal */
+export interface RequestPermissionDialogProps {
+  onClose?: () => void
+  onRequestSubmitted?: () => void
+}
+
+/**
+ * A confirmation dialog used to prevent unwanted document deletes. Loads all
+ * the referencing internal and cross-data references prior to showing the
+ * delete button.
+ *
+ * @internal
+ */
+export function RequestPermissionDialog({
+  onClose,
+  onRequestSubmitted,
+}: RequestPermissionDialogProps) {
+  const {t} = useTranslation(structureLocaleNamespace)
+  const telemtry = useTelemetry()
+  const dialogId = `request-permissions-${useId()}`
+  const projectId = useProjectId()
+  const client = useClient({apiVersion: '2024-09-26'})
+  const toast = useToast()
+  const zOffset = useZIndex()
+
+  const [isSubmitting, setIsSubmitting] = useState(false)
+
+  const [note, setNote] = useState('')
+  const [noteLength, setNoteLength] = useState<number>(0)
+
+  const [msgError, setMsgError] = useState<string | undefined>()
+  const [hasTooManyRequests, setHasTooManyRequests] = useState<boolean>(false)
+  const [hasBeenDenied, setHasBeenDenied] = useState<boolean>(false)
+
+  const requestedRole$: Observable<'administrator' | 'editor'> = useMemo(() => {
```

Comment:
Is there always an administrator role, even when using custom roles/role mapping? Or do we assume that if the customer has a `viewer` role they are using default roles and it shouldn't matter?

Reply:
Second one is our inference